crypto/o_fips.c

```diff
@@ -1,24 +1,62 @@
 /*
- * Copyright 2011-2016 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2011-2020 The OpenSSL Project Authors. All Rights Reserved.
  *
  * Licensed under the Apache License 2.0 (the "License").  You may not use
  * this file except in compliance with the License.  You can obtain a copy
  * in the file LICENSE in the source distribution or at
  * https://www.openssl.org/source/license.html
  */
 
+#include "crypto/evp.h"
 #include "internal/cryptlib.h"
 
+#define FIPS_QUERY_ON  "fips=yes"
+#define FIPS_QUERY_OFF ""
+#define DIGEST_NAME "SHA-256"
+
+/*
+ * Check that the library context's default properties has the value "fips=yes"
+ */
+int FIPS_mode_with_libctx(OPENSSL_CTX *libctx)
+{
+    return evp_default_property_is_enabled(libctx, "fips");
+}
+
 int FIPS_mode(void)
 {
-    /* This version of the library does not support FIPS mode. */
-    return 0;
+    return FIPS_mode_with_libctx(NULL);
+}
+
+/*
+ * Sets the default properties for a library context to either
+ * “fips=yes” if on != 0 OR
+ * “” for on == 0. i.e (fips is unset)
+ * If on != 0 it also fetches the SHA256 algorithm with the property
+ * “fips=yes” and ensure that it successfully returns.
+ */
+int FIPS_mode_set_with_libctx(OPENSSL_CTX *libctx, int on)
+{
+    int ret = 1;
+    int last_on = FIPS_mode_with_libctx(libctx);
+
+    if (!EVP_set_default_properties(libctx, on != 0 ? FIPS_QUERY_ON :
+                                                      FIPS_QUERY_OFF))
```

Comment:
I've just been discussing this with @levitte. It think we have a problem here. Setting the default properties like this is going to mean that we cannot load any serializers - because they aren't in the FIPS provider.

Reply:
This sounds like a fundamental design problem to me :)
If both the fips and default are always going to be present then how do we know its fetching the fips algorithm (which was the whole point of being able to specify it in the default properties).

Reply:
Yes - I agree there is a design issue here. It was always the plan that we would do it the way you're doing it in this PR. We just never appreciated the impact on serializers. @levitte is considering having a separate serializer provider - but this only goes part off the way to solving the problem. Probably an item to discuss on a call.

Reply:
So it should not be too hard to change this function to do a merge with other properties instead of a straight set like it is doing now.. i.e leave other default properties alone and just turn on or off the fip=yes setting.
That wont stop someone calling EVP_set_default_properties() directly though.

Reply:
> So it should not be too hard to change this function to do a merge with other properties instead of a straight set like it is doing now.. i.e leave other default properties alone and just turn on or off the fip=yes setting.

But that doesn't entirely help either since the serializer provider won't have "fips=yes", so the serializers will never match the query properties. You could have it do "?fips"=yes" instead. But that would mean that (for example) if you had the default provider loaded and attempted to fetch a non-fipsable algorithm, then it would work - which is undesirable.

Reply:
hmm will have to think about where the serializer stuff really belongs..

Reply:
Could we register the serialisers in the default provider including "fips=yes" in their properties?  The properties are specified per algorithm.

Reply:
That's possible but quite confusing. If we went that route, I'd suggest we put it in a "Serialisation" provider rather than the default provider (as mentioned above).

Reply:
I'd be fine with a serialisation provider too, it would still need to define _fips=yes_ for its implementations.

Reply:
> That's possible but quite confusing.

I think the thing that is confusing about this is that we have properties such as "default=yes", "legacy=yes" and "fips=yes", which until now have always meant "the algorithm comes from the default, legacy or fips providers". Would it be better to have a single property for this "provider=legacy", "provider=default", "provider=fips"? Then we could have "fips=yes" as a separate property meaning "the algorithm is compatible for use in FIPS mode". This would make things a lot clearer I think.

Reply:
And it scales better when providers are added in the future.

Reply:
I've made this change in #11097